src/Microsoft.ML.FastTree/GamModelParameters.cs

```diff
@@ -0,0 +1,911 @@
+            {
+                var featuresIndices = features.GetIndices();
+                // Add in the precomputed results for all features
+                value += _valueAtAllZero;
+                Array.Copy(_binsAtAllZero, bins, _numFeatures);
+
+                // Update the results for features we have
+                for (int i = 0; i < featuresValues.Length; ++i)
+                {
+                    if (_inputFeatureToDatasetFeatureMap.TryGetValue(featuresIndices[i], out int j))
+                        // Add the value and subtract the value at zero that was previously accounted for
+                        value += GetBinEffect(j, featuresValues[i], out bins[j]) - GetBinEffect(j, 0);
+                }
+            }
+            return value;
+        }
+
+        private double GetBinEffect(int featureIndex, double featureValue)
+        {
+            Contracts.Assert(0 <= featureIndex && featureIndex < _numFeatures);
+            int index = Algorithms.FindFirstGE(_binUpperBounds[featureIndex], featureValue);
+            return _binEffects[featureIndex][index];
+        }
+
+        private double GetBinEffect(int featureIndex, double featureValue, out int binIndex)
+        {
+            Contracts.Assert(0 <= featureIndex && featureIndex < _numFeatures);
+            binIndex = Algorithms.FindFirstGE(_binUpperBounds[featureIndex], featureValue);
+            return _binEffects[featureIndex][binIndex];
+        }
+
+        /// <summary>
+        /// Get the bin upper bounds for each feature.
+        /// </summary>
+        /// <param name="featureIndex">The index of the feature (in the training vector) to get.</param>
+        /// <returns>The bin upper bounds. May be null if this feature has no bins.</returns>
+        public double[] GetFeatureBinUpperBounds(int featureIndex)
+        {
+            Contracts.Assert(0 <= featureIndex && featureIndex < _numFeatures);
+            double[] featureBins;
+            if (_inputFeatureToDatasetFeatureMap.TryGetValue(featureIndex, out int j))
+            {
+                featureBins = new double[_binUpperBounds[j].Length];
+                _binUpperBounds[j].CopyTo(featureBins, 0);
+            }
+            else
+            {
+                featureBins = new double[0];
+            }
+
+            return featureBins;
+        }
+
+        /// <summary>
+        /// Get the binned weights for each feature.
+        /// </summary>
+        /// <param name="featureIndex">The index of the feature (in the training vector) to get.</param>
+        /// <returns>The binned weights for each feature.</returns>
+        public double[] GetFeatureWeights(int featureIndex)
+        {
+            Contracts.Assert(0 <= featureIndex && featureIndex < _numFeatures);
+            double[] featureWeights;
+            if (_inputFeatureToDatasetFeatureMap.TryGetValue(featureIndex, out int j))
+            {
+                featureWeights = new double[_binUpperBounds[j].Length];
```

Comment:
>_binUpperBounds [](start = 44, length = 15)

Just checking that this is the same length as _binEffects[j]

Reply:
Good catch! Should be the same and checked when defined, but not good practice. Changing to use _binEffects[j].Length.